Rich-Text-Editor-(RTE)-Overview.md

```diff
@@ -1,117 +1,137 @@
+
+Near the end of the `ngAfterViewInit()` function, we use `ck.on` to specify the following callback functions:
+
+* The function passed to `ck.on('instanceReady', ...)` is executed when any new instance of CKEditor is ready. For example it will be executed when we click on the content of a card, for each option that we have for multiple choice options etc. Here we can add the css and icons to the buttons in the toolbar.
+* The function passed to `ck.on('change', ...)` is executed whenever the content of CKEditor changes. It handles user updates to the rich text.
+
+### Adding a new plugin
+
+To add a new plugin, follow these steps:
+
+1. Add the plugin to the `manifest.json` to download it. `SharedSpace` can be looked as an example. Alternatively, define your own custom plugin like we did with `pre`.
+2. Add a call to `CKEditor.plugins.addExternal` to the component's `ngAfterViewInit()` method.
+3. Add this plugin to `extraPlugins` in `ckConfig`.
+
+#### Adding rich text components to CKEditor
+
+Rich text components are added to CKEditor in [`ck-editor-4-widgets.initializer.ts`](https://github.com/oppia/oppia/blob/develop/core/templates/components/ck-editor-helpers/ck-editor-4-widgets.initializer.ts). The components are dynamically added to CKEditor as [widgets](https://docs.ckeditor.com/ckeditor4/latest/guide/widget_sdk_intro.html). We use the `getRichTextComponents()` function to obtain the information of each rich text component and construct there respective widgets. We also have a function `isInlineComponent()` to check whether a rich text component is inline or block component.
+
+`componentTemplate` defines a template to wrap the rich text components. Inline components are wrapped in a span, and block components are wrapped in a div.
+
+The plugins are added to CKEditor in the line `CKEDITOR.plugins.add(ckName, {`. The `init` function passed to `add()` is executed when the plugin is initialised and it adds a widget for each component using the line `editor.widgets.add(ckName, {`. `editor.widgits.add` is passed an object with the following attributes:
+
+* `button`: The button label for the widget.
+* `inline`: Whether the widget is an inline component or block component. The link and the math component are inline and the rest are declared block components.
+* `template`: The wrapper template for the widget.
+* `edit`: A function that will be executed when a widget is being edited. In this function the default action is canceled since we have used our own edit modal. `RteHelperService._openCustomizationModal()`is a helper function for opening the modal which is used to insert new components or editing existing ones. It uses the customizationArgSpecs (obtained from the component definition) to know what the editable fields are for each component, which allows it to render the modal properly for any component.
```

Comment:
Should customizationArgSpecs link to an explanation of what these are?

Reply:
done